lms/static/scripts/frontend_apps/components/ErrorDisplay.js

```diff
@@ -30,82 +30,114 @@ function toSentence(str) {
  */
 
 /**
- * @typedef ErrorDisplayProps
- * @prop {string|null} [message] -
- *   A short message to display explaining that a problem happened. This is
- *   typically a general message like "There was a problem fetching this assignment".
- *   In cases where the the error originates from the server, this message may not
- *   be necessary, but in other cases where the `error.message` is generic and perhaps
- *   originates from an exception in the client, then this prop can be used to
- *   provide additional specifics.
- * @prop {ErrorLike} error -
- *   An `Error`-like object with specific details of the problem. If `error` contains
- *   a `message` property, then that string will be rendered.
- */
-
-/**
- * Displays details of an error, such as a failed API call and provide the user
- * with information on how to get help with it.
+ * JSON-stringify `error.details` if it is extant and an object
  *
- * @param {ErrorDisplayProps} props
+ * @param {ErrorLike} error
  */
-export default function ErrorDisplay({ message, error }) {
-  /** @type {string|undefined} */
-  let details = '';
-
-  if (typeof error.details === 'object' && error.details !== null) {
+function formatErrorDetails(error) {
+  /** @type {string|object} */
+  let details = error.details ?? '';
+  if (error?.details && typeof error.details === 'object') {
     try {
       details = JSON.stringify(error.details, null, 2 /* indent */);
     } catch (e) {
       // ignore
     }
-  } else {
-    details = error.details;
   }
+  return details;
+}
 
-  let supportEmailBody = '';
-  if (error.message) {
-    supportEmailBody += `\n\nError message: ${error.message}`;
-  }
-  if (details) {
-    supportEmailBody += `\n\nTechnical details:\n\n${details}`;
+/**
+ * @typedef ErrorDetailsProps
+ * @prop {ErrorLike} error
+ *
+ * Render pre-formatted JSON details of an error
+ *
+ * @param {ErrorDetailsProps} props
+ */
+function ErrorDetails({ error }) {
+  /** @param {Event} event */
+  const onDetailsToggle = event => {
+    const detailsEl = /** @type {HTMLDetailsElement} */ (event.target);
+    if (!detailsEl.open) {
+      return;
+    }
+    detailsEl.scrollIntoView({ behavior: 'smooth' });
+  };
+
+  const details = formatErrorDetails(error);
+  if (!details) {
+    return null;
   }
 
+  return (
+    <details className="ErrorDisplay__details" onToggle={onDetailsToggle}>
+      <summary className="ErrorDisplay__details-summary">Error Details</summary>
+      <pre className="ErrorDisplay__details-content">{details}</pre>
+    </details>
+  );
+}
+
+/**
+ * @typedef ErrorDisplayProps
+ * @prop {string|null} [description] -
+ *   A short message explaining the error and its human-facing relevance.
+ *   This is typically a general message like
+ *   "There was a problem fetching this assignment". This description
+ *   always comes from (this) client app.
```

Comment:
Is this correct?

```suggestion
 *   always comes from the LMS app, and should be used only when the `error` prop contains insufficient information.
```

Reply:
This isn't _quite_ accurate. The `description` should only be set when the nature/type of the error isn't known (the error is being sort of "passed through", as it were). I seem to be having trouble articulating this; I'd like to make it better.

There are two general contexts here:

1. We know what kind of error the user is looking at, and we know how to explain it to them. e.g. "You've reused a consumer key" or "there aren't any files in this folder; you need to upload some." For those, we have some prepared text we can show them, external to this component. Error data details (as shown in the "Details" summary disclosure) _might_ be helpful for further debug. They're collapsed by default to stay out of the way. However, printing another error _message_ is not useful.
2. We may know the general category of error (A VitalSource API request has failed, e.g.) but not the specifics of it (e.g. "That book doesn't exist/404"). Thus we set the `description` to fulfill the general category ("Something went wrong when trying to get that book") and "pass" the error message through and show it.

Does this make general sense? I'll try to find a better way to explain it in comments.

Reply:
It does, thanks.